src/smoosh/src/smoosh_channel.erl

```diff
@@ -194,18 +236,122 @@ handle_info(check_window, State0) ->
+handle_info(activate, State) ->
+    {noreply, activate_channel(State)};
+handle_info(persist, State) ->
+    persist_queue(State),
+    erlang:send_after(?CHECKPOINT_INTERVAL_IN_MSEC, self(), persist),
+    {noreply, State};
+handle_info(pause, State) ->
     {noreply, State#state{paused = true}};
-handle_info(unpause, State0) ->
-    {ok, State} = code_change(nil, State0, nil),
+handle_info(unpause, State) ->
     {noreply, maybe_start_compaction(State#state{paused = false})}.
 
 terminate(_Reason, _State) ->
     ok.
 
-code_change(_OldVsn, #state{} = State, _Extra) ->
-    {ok, State}.
+persist_queue(#state{waiting = Waiting} = State) ->
+    write_state_to_file(State),
+    smoosh_priority_queue:write_to_file(Waiting).
+
+recover(FilePath) ->
+    case do_recover(FilePath) of
+        {ok, List} ->
+            List;
+        error ->
+            []
+    end.
+
+do_recover(FilePath) ->
+    case file:read_file(FilePath) of
+        {ok, Content} ->
+            <<Vsn, Binary/binary>> = Content,
+            try parse_state(Vsn, ?VSN, Binary) of
+                Term ->
+                    couch_log:notice(
+                        "~p Successfully restored state file ~s", [?MODULE, FilePath]
+                    ),
+                    {ok, Term}
+            catch
+                error:Reason ->
+                    couch_log:error(
+                        "~p Invalid state file (~p). Deleting ~s", [?MODULE, Reason, FilePath]
+                    ),
+                    file:delete(FilePath),
+                    error
+            end;
+        {error, enoent} ->
+            couch_log:notice(
+                "~p (~p) State file ~s does not exist. Not restoring.", [?MODULE, enoent, FilePath]
+            ),
+            error;
+        {error, Reason} ->
+            couch_log:error(
+                "~p Cannot read the state file (~p). Deleting ~s", [?MODULE, Reason, FilePath]
+            ),
+            file:delete(FilePath),
+            error
+    end.
+
+parse_state(1, ?VSN, Binary) ->
+    erlang:binary_to_term(Binary, [safe]);
+parse_state(Vsn, ?VSN, _) ->
+    error({unsupported_version, Vsn}).
+
+write_state_to_file(#state{name = Name, active = Active, starting = Starting}) ->
+    Active1 = lists:foldl(
+        fun({DbName, _}, Acc) ->
+            [DbName | Acc]
+        end,
+        [],
+        Active
+    ),
+    Starting1 = lists:foldl(
+        fun({_, DbName}, Acc) ->
+            [DbName | Acc]
+        end,
+        [],
+        Starting
+    ),
+    write_to_file(Active1, active_file_name(Name)),
+    write_to_file(Starting1, starting_file_name(Name)).
+
+write_to_file(List, FileName) ->
+    couch_log:notice("~p Writing state to state file ~s", [?MODULE, FileName]),
+    OnDisk = <<?VSN, (erlang:term_to_binary(List, [compressed, {minor_version, 1}]))/binary>>,
+    TmpFileName = FileName ++ ".tmp",
+    file:delete(TmpFileName),
+    file:write_file(TmpFileName, OnDisk, [sync]),
```

Comment:
I think we might want to handle errors better here. We don't want to call `file:delete(FileName)` if `file:write_file` return `{error, Reason}`. We might also want to log this case.

Reply:
or use `ok = ...` idiom and let standard error logging to handle the case.